libraries/script-engine/src/ScriptEngine.cpp

```diff
@@ -135,38 +135,53 @@ QString encodeEntityIdIntoEntityUrl(const QString& url, const QString& entityID)
     return url + " [EntityID:" + entityID + "]";
 }
 
-static bool hasCorrectSyntax(const QScriptProgram& program, ScriptEngine* reportingEngine) {
-    const auto syntaxCheck = QScriptEngine::checkSyntax(program.sourceCode());
-    if (syntaxCheck.state() != QScriptSyntaxCheckResult::Valid) {
+QScriptProgram ScriptEngine::compileScript(const QString& sourceCode, const QString& fileName, const int lineNumber) {
+    QScriptProgram program(sourceCode, fileName, lineNumber);
+    const auto syntaxCheck = checkSyntax(program.sourceCode());
+    if (syntaxCheck.state() != syntaxCheck.Valid) {
         const auto error = syntaxCheck.errorMessage();
         const auto line = QString::number(syntaxCheck.errorLineNumber());
         const auto column = QString::number(syntaxCheck.errorColumnNumber());
         const auto message = QString("[SyntaxError] %1 in %2:%3(%4)").arg(error, program.fileName(), line, column);
-        reportingEngine->scriptErrorMessage(qPrintable(message));
-        return false;
+        scriptErrorMessage(qPrintable(message));
+        return QScriptProgram();
     }
-    return true;
+    return program;
 }
 
-static bool hadUncaughtExceptions(QScriptEngine& engine, const QString& fileName, ScriptEngine* reportingEngine, QString* exceptionMessage = nullptr) {
-    if (engine.hasUncaughtException()) {
-        const auto backtrace = engine.uncaughtExceptionBacktrace();
-        const auto exception = engine.uncaughtException().toString();
-        const auto line = QString::number(engine.uncaughtExceptionLineNumber());
-        engine.clearExceptions();
+QString ScriptEngine::reportUncaughtExceptions(const QString& fileName, QScriptEngine* engine) {
+    if (engine == nullptr) {
+        engine = this;
+    }
+    if (hasUncaughtException()) {
+        const auto backtrace = engine->uncaughtExceptionBacktrace();
+        const auto exception = engine->uncaughtException().toString();
+        const auto line = QString::number(engine->uncaughtExceptionLineNumber());
+        const auto nestedEval = engine == this && evaluatePending();
+        if (!nestedEval) {
+            // consume the exception only if the top-level evaluate/evaluateInClosure
+            engine->clearExceptions();
+        }
 
         QString message = QString(SCRIPT_EXCEPTION_FORMAT).arg(exception, fileName, line);
         if (!backtrace.empty()) {
             static const auto lineSeparator = "\n    ";
             message += QString("\n[Backtrace]%1%2").arg(lineSeparator, backtrace.join(lineSeparator));
         }
-        reportingEngine->scriptErrorMessage(qPrintable(message));
```

Comment:
`reportingEngine` was being used to convey a copy of the `this` pointer (the function used to be a free standing static method -- but is now an instance method, so the indirect pointer no longer needed).